### PR TITLE
Feat/crm deal stage apis for edit delete and reordering

### PR DESCRIPTION
**Pull request by sanjula-rtc**
### Summary

-

### How to test

1.

### Project Checklist

- [x] Changes build without any errors
- [x] Have written adequate test cases
- [x] Done developer testing in
  - [ ] Chrome
  - [ ] Firefox
  - [ ] Safari
- [ ] Code is formatted with `npm run format`
- [ ] Code is linted with `npm run check-lint`
- [ ] No unnecessary comments left in code
- [ ] Made corresponding changes to the documentation

#### Other

- [ ] New atomic components added
- [ ] New molecules added
- [ ] New pages(routes) added
- [ ] New dependencies installed

### PR Checklist

- [ ] Pull request is raised from the correct source branch
- [ ] Pull request is raised to the correct destination branch
- [ ] Pull request is raised with correct title
- [ ] Pull request is self reviewed
- [ ] Pull request is self assigned
- [ ] Suitable pull request status labels are added (`ready-for-code-review`)

### Additional Information
-

### CODE REVIEW

File: backend/src/main/java/com/skapp/community/crmplanner/controller/v1/CrmDealStageController.java [hunk 1]
```diff
@@ -26,7 +33,7 @@ public class CrmDealStageController {
 
 	@Operation(summary = "Get deal stages", description = "Returns all active deal stages ordered by index.")
 	@GetMapping(produces = MediaType.APPLICATION_JSON_VALUE)
-	@PreAuthorize("hasAnyRole('ROLE_CRM_SALES_REPRESENTATIVE')")
+	@PreAuthorize("hasAnyRole('ROLE_CRM_ADMIN')")
 	public ResponseEntity<ResponseEntityDto> getDealStages() {
```
[the remark this answers is not in the document]
Comment (sanjula-rtc) [reply backend/src/main/java/com/skapp/community/crmplanner/controller/v1/CrmDealStageController.java:37]: fixed

File: backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java [hunk 2]
```diff
@@ -68,6 +77,77 @@ public ResponseEntityDto createDealStage(CrmDealStageCreateRequestDto requestDto
...
+	@Transactional
+	public ResponseEntityDto deleteDealStage(Long id) {
+		log.info("deleteDealStage: execution started for id={}", id);
+
+		CrmDealStage stage = crmDealStageDao.findByIdAndIsDeletedFalse(id)
+			.orElseThrow(() -> new ModuleException(CrmMessageConstant.CRM_ERROR_DEAL_STAGE_NOT_FOUND));
+
+		stage.setIsDeleted(true);
+		crmDealStageDao.save(stage);
+
```
[the remark this answers is not in the document]
Comment (sanjula-rtc) [reply backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java:136]: fixed

File: backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java [hunk 3]
```diff
@@ -68,6 +77,77 @@ public ResponseEntityDto createDealStage(CrmDealStageCreateRequestDto requestDto
...
+		List<CrmDealStage> existingStages = crmDealStageDao.findAllByIdInAndIsDeletedFalse(stageIds);
+
+		if (existingStages.size() != stageIds.size()) {
+			throw new ModuleException(CrmMessageConstant.CRM_ERROR_DEAL_STAGE_NOT_FOUND);
+		}
+
+		log.info("reorderDealStages: execution ended, reordered {} stages", stages.size());
+
+		return new ResponseEntityDto(false, crmMapper.crmDealStagesToCrmDealStageResponseDtos(existingStages));
+	}
```
[the remark this answers is not in the document]
Comment (sanjula-rtc) [reply backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java:179]: fixed

File: backend/src/main/java/com/skapp/community/crmplanner/util/CrmValidations.java [hunk 4]
```diff
@@ -13,13 +13,18 @@
...
 
 import java.net.MalformedURLException;
 import java.net.URI;
 import java.net.URISyntaxException;
 import java.time.LocalDate;
 import java.time.LocalDateTime;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
```
[the remark this answers is not in the document]
Comment (sanjula-rtc) [reply backend/src/main/java/com/skapp/community/crmplanner/util/CrmValidations.java]: fixed

File: backend/src/main/java/com/skapp/community/crmplanner/controller/v1/CrmDealStageController.java [hunk 5]
```diff
@@ -40,4 +47,31 @@ public ResponseEntity<ResponseEntityDto> createDealStage(@RequestBody CrmDealSta
 		return new ResponseEntity<>(response, HttpStatus.CREATED);
 	}
 
+	@Operation(summary = "Edit a deal stage", description = "Updates a deal stage. All fields are optional.")
+	@PatchMapping("/{id}")
+	@PreAuthorize("hasAnyRole('ROLE_CRM_ADMIN')")
+	public ResponseEntity<ResponseEntityDto> editDealStage(@PathVariable Long id,
+			@RequestBody CrmDealStageEditRequestDto requestDto) {
```
[the remark this answers is not in the document]
Comment (sanjula-rtc) [reply backend/src/main/java/com/skapp/community/crmplanner/controller/v1/CrmDealStageController.java:54]: added

File: backend/src/main/java/com/skapp/community/crmplanner/util/CrmValidations.java [hunk 6]
```diff
@@ -298,4 +303,22 @@ public static void validateDomain(String domain) {
 		}
 	}
 
+	public static void validateDealStageReorderRequest(List<CrmDealStageReorderRequestDto> stages) {
+		if (stages == null || stages.isEmpty()) {
+			throw new ModuleException(CrmMessageConstant.CRM_ERROR_DEAL_STAGE_NOT_FOUND);
+		}
```
[the remark this answers is not in the document]
Comment (sanjula-rtc) [reply backend/src/main/java/com/skapp/community/crmplanner/util/CrmValidations.java:308]: it means deal stages are not found in the request

File: backend/src/main/resources/community/messages/crm-messages.properties [hunk 7]
```diff
@@ -80,5 +80,7 @@ api.error.crm.deal-stage.name-duplicate=A deal stage with this name already exis
 api.error.crm.deal-stage.description-too-long=Deal stage description must not exceed 250 characters
 api.error.crm.deal-stage.description-invalid-chars=Deal stage description contains invalid characters
 api.error.crm.deal-stage.color-required=Deal stage color is required
+api.error.crm.deal-stage.reorder-invalid-request=Deal stage reorder request contains null ID or order index
+api.error.crm.deal-stage.reorder-duplicate-values=Deal stage reorder request contains duplicate IDs or order indexes
 api.success.crm.deal-deleted=Deal deleted successfully
 api.success.crm.task-deleted=Task deleted successfully
```
[the remark this answers is not in the document]
Comment (sanjula-rtc) [reply backend/src/main/resources/community/messages/crm-messages.properties]: done

**Review by ThinuwanW — COMMENTED** [review]
## 🤖 Claude Code Review

This PR adds edit, soft-delete, and reorder endpoints for CRM deal stages, following existing layered conventions (controller→service→DAO, manual validation via CrmValidations, ResponseEntityDto). Code structure, security annotations, logging, and tests are largely consistent. However, two data-integrity gaps stand out: the reorder logic can create multiple INITIAL stages / duplicate order indexes on a partial request, and stage deletion does not check for deals still referencing the stage (a non-null FK), which orphans those deals.

Found **10** new issue(s): 🔴 2 important, 🟡 2 suggestion(s), 🟣 2 nit(s)

File: backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java [hunk 8]
```diff
@@ -68,6 +78,104 @@ public ResponseEntityDto createDealStage(CrmDealStageCreateRequestDto requestDto
...
+
+		Map<Long, CrmDealStage> exisitingStagesMap = existingStages.stream()
+			.collect(Collectors.toMap(CrmDealStage::getId, stage -> stage));
+
+		changedStages.forEach(newStage -> {
+			CrmDealStage stage = exisitingStagesMap.get(newStage.getId());
+			stage.setOrderIndex(newStage.getOrderIndex());
+		});
+
+		ensureFirstStageIsInitial(existingStages);
```
Comment (ThinuwanW) [inline backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java]: 🤖 **Claude** · 🔴 **Important**

reorderDealStages/ensureFirstStageIsInitial only operate on the stages included in the request. It promotes the minimum-orderIndex stage in the SUBSET to INITIAL and demotes the rest to OPEN, but nothing forces the request to contain ALL non-terminal stages, and INITIAL is not in TERMINAL_STAGES so it is not required to be present. If the caller omits the current INITIAL stage (or sends only OPEN stages), the existing INITIAL stage keeps its INITIAL type while a subset stage is also set to INITIAL — resulting in TWO INITIAL stages and likely duplicate global orderIndex values (the new INITIAL gets orderIndex 1 while the old INITIAL still has 1). The 'valid' integration test reorderDealStages_ValidOpenStages_ReturnsOkAndUpdatesInitial reproduces exactly this (it reorders only the 4 OPEN stages, never including LEAD/INITIAL) and only asserts HTTP 200, so the corrupted state passes silently. Fix: either validate that the request contains every active non-terminal stage before reordering, or load all non-terminal stages, apply the new indexes, and recompute INITIAL/OPEN across the full set so the previous INITIAL is demoted.

File: backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java [hunk 9]
```diff
@@ -68,6 +78,104 @@ public ResponseEntityDto createDealStage(CrmDealStageCreateRequestDto requestDto
...
+		log.info("deleteDealStage: execution started for id={}", id);
+
+		CrmDealStage stage = crmDealStageDao.findByIdAndIsDeletedFalse(id)
+			.orElseThrow(() -> new ModuleException(CrmMessageConstant.CRM_ERROR_DEAL_STAGE_NOT_FOUND));
+
+		if (CrmConstants.NON_DELETEABLE_STAGES.contains(stage.getStageType())) {
+			throw new ModuleException(CrmMessageConstant.CRM_ERROR_CANNOT_DELETE_TERMINAL_STAGE);
+		}
+
+		stage.setIsDeleted(true);
```
Comment (ThinuwanW) [inline backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java:134]: 🤖 **Claude** · 🔴 **Important**

deleteDealStage soft-deletes a stage after only checking NON_DELETEABLE_STAGES, but does not verify whether any deals still reference it. CrmDeal.stage is mapped with @JoinColumn(name="stage_id", nullable=false), so deals in a deleted OPEN stage become orphaned: getDealStages filters isDeletedFalse so the stage disappears from the board, yet the deals still point to it (their non-null FK now references an invisible stage). Add a guard that rejects deletion when active deals reference the stage (e.g. a countByStage_IdAndIsDeletedFalse check returning a dedicated error message), or require the caller to migrate those deals first.

File: backend/src/main/java/com/skapp/community/crmplanner/constant/CrmMessageConstant.java [hunk 10]
```diff
@@ -69,9 +69,13 @@ public enum CrmMessageConstant implements MessageConstant {
 	CRM_ERROR_DEAL_STAGE_NAME_LENGTH("api.error.crm.deal-stage.name-length"),
 	CRM_ERROR_DEAL_STAGE_NAME_INVALID_CHARS("api.error.crm.deal-stage.name-invalid-chars"),
 	CRM_ERROR_DEAL_STAGE_NAME_DUPLICATE("api.error.crm.deal-stage.name-duplicate"),
+	CRM_SUCCESS_DEAL_STAGE_DELETED("api.success.crm.deal-stage-deleted"),
 	CRM_ERROR_DEAL_STAGE_DESCRIPTION_TOO_LONG("api.error.crm.deal-stage.description-too-long"),
 	CRM_ERROR_DEAL_STAGE_DESCRIPTION_INVALID_CHARS("api.error.crm.deal-stage.description-invalid-chars"),
 	CRM_ERROR_DEAL_STAGE_COLOR_REQUIRED("api.error.crm.deal-stage.color-required"),
+	CRM_ERROR_INVALID_REQUEST("api.error.crm.deal-stage.reorder-invalid-request"),
+	CRM_ERROR_INVALID_ORDER_INDEX("api.error.crm.deal-stage.reorder-invalid-order-index"),
```
Comment (ThinuwanW) [inline backend/src/main/java/com/skapp/community/crmplanner/constant/CrmMessageConstant.java]: 🤖 **Claude** · 🟡 **Suggestion**

CRM_ERROR_INVALID_ORDER_INDEX (key api.error.crm.deal-stage.reorder-invalid-order-index) is never referenced — validateDealStageReorderRequest uses CRM_ERROR_INVALID_REQUEST for both null id and null orderIndex. This constant and its corresponding property in crm-messages.properties are dead code. Remove both, or use this constant for the null-orderIndex branch so the message matches the failure.

File: backend/src/main/java/com/skapp/community/crmplanner/util/CrmValidations.java [hunk 11]
```diff
@@ -298,4 +303,22 @@ public static void validateDomain(String domain) {
 		}
 	}
 
+	public static void validateDealStageReorderRequest(List<CrmDealStageReorderRequestDto> stages) {
+		if (stages == null || stages.isEmpty()) {
+			throw new ModuleException(CrmMessageConstant.CRM_ERROR_DEAL_STAGE_NOT_FOUND);
```
Comment (ThinuwanW) [inline backend/src/main/java/com/skapp/community/crmplanner/util/CrmValidations.java]: 🤖 **Claude** · 🟡 **Suggestion**

validateDealStageReorderRequest throws CRM_ERROR_DEAL_STAGE_NOT_FOUND when the list is null or empty. 'Deal stage not found' is semantically wrong for an empty/missing request body — it should signal an invalid request (e.g. CRM_ERROR_INVALID_REQUEST or a dedicated 'reorder list empty' message). The integration test asserts the current (misleading) message, so update both together.

File: backend/src/main/java/com/skapp/community/crmplanner/repository/CrmDealStageDao.java [hunk 12]
```diff
@@ -1,11 +1,13 @@
 package com.skapp.community.crmplanner.repository;
 
 import com.skapp.community.crmplanner.model.CrmDealStage;
+import com.skapp.community.crmplanner.type.CrmDealStageType;
```
Comment (ThinuwanW) [inline backend/src/main/java/com/skapp/community/crmplanner/repository/CrmDealStageDao.java:4]: 🤖 **Claude** · 🟣 **Nit**

The added imports CrmDealStageType (line 4) and java.util.Set (line 10) are not used anywhere in this interface. Unused imports will fail the Checkstyle/build gate. Remove them.

File: backend/src/main/java/com/skapp/community/crmplanner/constant/CrmMessageConstant.java [hunk 13]
```diff
@@ -69,9 +69,13 @@ public enum CrmMessageConstant implements MessageConstant {
 	CRM_ERROR_DEAL_STAGE_NAME_LENGTH("api.error.crm.deal-stage.name-length"),
 	CRM_ERROR_DEAL_STAGE_NAME_INVALID_CHARS("api.error.crm.deal-stage.name-invalid-chars"),
 	CRM_ERROR_DEAL_STAGE_NAME_DUPLICATE("api.error.crm.deal-stage.name-duplicate"),
+	CRM_SUCCESS_DEAL_STAGE_DELETED("api.success.crm.deal-stage-deleted"),
 	CRM_ERROR_DEAL_STAGE_DESCRIPTION_TOO_LONG("api.error.crm.deal-stage.description-too-long"),
 	CRM_ERROR_DEAL_STAGE_DESCRIPTION_INVALID_CHARS("api.error.crm.deal-stage.description-invalid-chars"),
 	CRM_ERROR_DEAL_STAGE_COLOR_REQUIRED("api.error.crm.deal-stage.color-required"),
+	CRM_ERROR_INVALID_REQUEST("api.error.crm.deal-stage.reorder-invalid-request"),
```
Comment (ThinuwanW) [inline backend/src/main/java/com/skapp/community/crmplanner/constant/CrmMessageConstant.java]: 🤖 **Claude** · 🟣 **Nit**

The new constant names CRM_ERROR_INVALID_REQUEST and CRM_ERROR_DUPLICATE_VALUES are generic, but their keys are reorder-specific (reorder-invalid-request, reorder-duplicate-values) and they are only used by the reorder flow. Rename to reorder-scoped names (e.g. CRM_ERROR_DEAL_STAGE_REORDER_INVALID_REQUEST / _DUPLICATE_VALUES) to match the naming of the surrounding deal-stage constants and avoid accidental reuse for unrelated errors.

File: backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java [hunk 14]
```diff
@@ -68,6 +78,104 @@ public ResponseEntityDto createDealStage(CrmDealStageCreateRequestDto requestDto
 		return new ResponseEntityDto(false, crmMapper.crmDealStageToCrmDealStageResponseDto(saved));
 	}
 
+	@Override
+	@Transactional
+	public ResponseEntityDto editDealStage(Long id, CrmDealStageEditRequestDto requestDto) {
```
Comment (ThinuwanW) [inline backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java:87]: 🤖 **Claude** · 💬 **Note**

Inconsistent business-rule protection across the new operations. deleteDealStage blocks NON_DELETEABLE_STAGES (INITIAL, WON, LOST) and reorderDealStages blocks TERMINAL_STAGES (WON, LOST), but editDealStage applies no such guard — an admin can rename and recolor the INITIAL/WON/LOST stages freely. If terminal/initial stages are meant to be protected (as delete/reorder imply), edit should reject editing them (or at least disallow name changes) with an appropriate ModuleException. If renaming labels is intended, please confirm and ignore. Flagging so the asymmetry is a deliberate decision.

File: backend/src/test/java/com/skapp/community/crmplanner/controller/v1/CrmDealStageControllerIntegrationTest.java [hunk 15]
```diff
@@ -300,17 +304,233 @@ void createDealStage_TooLongDescription_ReturnsBadRequest() throws Exception {
 				.value(messageUtil.getMessage(CrmMessageConstant.CRM_ERROR_DEAL_STAGE_DESCRIPTION_TOO_LONG)));
 	}
 
+	private ResultActions performDeleteRequest(Long id) throws Exception {
+		return performRequest(delete(BASE_PATH + "/" + id).accept(MediaType.APPLICATION_JSON));
+	}
+
+	private ResultActions performReorderRequest(List<CrmDealStageReorderRequestDto> payload) throws Exception {
```
Comment (ThinuwanW) [inline backend/src/test/java/com/skapp/community/crmplanner/controller/v1/CrmDealStageControllerIntegrationTest.java:312]: 🤖 **Claude** · 💬 **Note**

reorderDealStages_ValidOpenStages_ReturnsOkAndUpdatesInitial only asserts HTTP 200 / STATUS_SUCCESSFUL — it never verifies the core documented behavior (the minimum-orderIndex stage becomes INITIAL and others become OPEN). Because the first-pass bug about ensureFirstStageIsInitial operating only on the request subset would still pass this test, the test gives false confidence. Add assertions on the response body / a DB re-read confirming which stage's stageType is INITIAL after reorder.

File: backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java [hunk 16]
```diff
@@ -68,6 +78,104 @@ public ResponseEntityDto createDealStage(CrmDealStageCreateRequestDto requestDto
...
+
+		if (existingStages.size() != stageIds.size()) {
+			throw new ModuleException(CrmMessageConstant.CRM_ERROR_DEAL_STAGE_NOT_FOUND);
+		}
+
+		if (existingStages.stream().map(CrmDealStage::getStageType).anyMatch(CrmConstants.TERMINAL_STAGES::contains)) {
+			throw new ModuleException(CrmMessageConstant.CRM_ERROR_CANNOT_REORDER_TERMINAL_STAGE);
+		}
+
+		Map<Long, CrmDealStage> exisitingStagesMap = existingStages.stream()
```
Comment (ThinuwanW) [inline backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java]: 🤖 **Claude** · 💬 **Note**

Typo in local variable name: 'exisitingStagesMap' (lines 155 and 159) should be 'existingStagesMap'. Minor, but it reads awkwardly and is easy to fix while the code is fresh.

File: backend/src/main/java/com/skapp/community/crmplanner/constant/CrmMessageConstant.java [hunk 17]
```diff
@@ -69,9 +69,13 @@ public enum CrmMessageConstant implements MessageConstant {
 	CRM_ERROR_DEAL_STAGE_NAME_LENGTH("api.error.crm.deal-stage.name-length"),
 	CRM_ERROR_DEAL_STAGE_NAME_INVALID_CHARS("api.error.crm.deal-stage.name-invalid-chars"),
 	CRM_ERROR_DEAL_STAGE_NAME_DUPLICATE("api.error.crm.deal-stage.name-duplicate"),
+	CRM_SUCCESS_DEAL_STAGE_DELETED("api.success.crm.deal-stage-deleted"),
```
Comment (ThinuwanW) [inline backend/src/main/java/com/skapp/community/crmplanner/constant/CrmMessageConstant.java]: 🤖 **Claude** · 💬 **Note**

CRM_SUCCESS_DEAL_STAGE_DELETED is declared in the middle of the deal-stage ERROR constants block (between name-duplicate and description-too-long) rather than grouped with the other CRM_SUCCESS_* entries near the bottom (e.g. CRM_SUCCESS_DEAL_DELETED). The corresponding properties entry is correctly placed under the success section, so this is purely an ordering/readability nit in the enum.

**Review by ThinuwanW — COMMENTED** [review]
## 🤖 Claude Code Review

This PR adds edit, soft-delete, and reorder endpoints for CRM deal stages with corresponding service logic, validation, message constants, and integration tests. The endpoints are correctly secured with ROLE_CRM_ADMIN and follow the layered architecture, but the reorder logic has a correctness bug around the single-INITIAL invariant when a partial stage set is submitted, and delete does not guard against stages that still have deals assigned. A few minor dead-code and naming issues exist.

Found **4** new issue(s): 
_8 finding(s) already posted in a previous review._

File: backend/src/test/java/com/skapp/community/crmplanner/controller/v1/CrmDealStageControllerIntegrationTest.java [hunk 18]
```diff
@@ -7,6 +7,9 @@
 import com.skapp.community.crmplanner.constant.CrmMessageConstant;
 import com.skapp.community.crmplanner.constant.DefaultCrmDealStageTemplate;
 import com.skapp.community.crmplanner.payload.request.CrmDealStageCreateRequestDto;
+import com.skapp.community.crmplanner.payload.request.CrmDealStageReorderRequestDto;
```
Comment (ThinuwanW) [inline backend/src/test/java/com/skapp/community/crmplanner/controller/v1/CrmDealStageControllerIntegrationTest.java:10]: 🤖 **Claude** · 💬 **Note**

📁 **File-level observation**

Missing test coverage: the new editDealStage (PATCH /v1/crm/deal/stage/{id}) endpoint has NO integration tests, while deleteDealStage and reorderDealStages each got thorough coverage (success, validation, terminal-stage, and forbidden cases). editDealStage has non-trivial branching (name-change duplicate check, optional description/color validation, not-found path, forbidden role) that is entirely untested. Add tests mirroring the delete/reorder suites: edit name/description/color happy path, duplicate-name rejection, invalid name/description/color, non-existent id (CRM_ERROR_DEAL_STAGE_NOT_FOUND), and a without-admin-role forbidden case.

File: backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java [hunk 19]
```diff
@@ -68,6 +78,104 @@ public ResponseEntityDto createDealStage(CrmDealStageCreateRequestDto requestDto
...
+			stage.setOrderIndex(newStage.getOrderIndex());
+		});
+
+		ensureFirstStageIsInitial(existingStages);
+
+		crmDealStageDao.saveAll(existingStages);
+
+		log.info("reorderDealStages: execution ended");
+
+		return new ResponseEntityDto(false, crmMapper.crmDealStagesToCrmDealStageResponseDtos(existingStages));
```
Comment (ThinuwanW) [inline backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java:178]: 🤖 **Claude** · 💬 **Note**

reorderDealStages returns crmMapper.crmDealStagesToCrmDealStageResponseDtos(existingStages), but existingStages comes from findAllByIdInAndIsDeletedFalse (no ORDER BY) and is never re-sorted, so the response list is in arbitrary DB order — unlike getDealStages which returns stages ordered by orderIndex. It is also only the subset that was passed in, not the full stage list. A client expecting an ordered, complete list (consistent with GET) will misrender. Sort by orderIndex before mapping (e.g. existingStages.sort(Comparator.comparing(CrmDealStage::getOrderIndex))) or return the full ordered set via findAllByIsDeletedFalseOrderByOrderIndexAsc().

File: backend/src/main/java/com/skapp/community/crmplanner/payload/request/CrmDealStageEditRequestDto.java [hunk 20]
```diff
@@ -0,0 +1,17 @@
+package com.skapp.community.crmplanner.payload.request;
```
Comment (ThinuwanW) [inline backend/src/main/java/com/skapp/community/crmplanner/payload/request/CrmDealStageEditRequestDto.java:1]: 🤖 **Claude** · 💬 **Note**

📁 **File-level observation**

Duplication: CrmDealStageEditRequestDto declares exactly the same three fields (name, description, color) with the same types as CrmDealStageCreateRequestDto. Per the duplication guideline, consider consolidating — e.g. have the edit DTO extend a shared base, or reuse the create DTO — to avoid two near-identical request payloads drifting apart over time. (Their semantics differ only in field optionality, which is enforced in the service, not the DTO.)

File: backend/src/main/java/com/skapp/community/crmplanner/controller/v1/CrmDealStageController.java [hunk 21]
```diff
@@ -40,4 +47,31 @@ public ResponseEntity<ResponseEntityDto> createDealStage(@RequestBody CrmDealSta
...
+
+	@Operation(summary = "Delete a deal stage", description = "Soft-deletes a deal stage.")
+	@DeleteMapping("/{id}")
+	@PreAuthorize("hasAnyRole('ROLE_CRM_ADMIN')")
+	public ResponseEntity<ResponseEntityDto> deleteDealStage(@PathVariable Long id) {
+		ResponseEntityDto response = crmDealStageService.deleteDealStage(id);
+		return new ResponseEntity<>(response, HttpStatus.OK);
+	}
+
+	@Operation(summary = "Reorder deal stages",
```
Comment (ThinuwanW) [inline backend/src/main/java/com/skapp/community/crmplanner/controller/v1/CrmDealStageController.java]: 🤖 **Claude** · 💬 **Note**

Misleading API documentation: the @Operation description states 'Reorders all deal stages. The stage with minimum orderIndex will be set to INITIAL type.' but the implementation only updates the subset of stages present in the request body and recomputes INITIAL within that subset (see related first-pass finding on the single-INITIAL invariant). Update the Swagger description to accurately reflect that only the provided stages are reordered, or change the implementation to operate on the full set.

File: backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java (hunk 14 above)
Comment (sanjula-rtc) [reply backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java:87]: it's inteded

File: backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java (hunk 9 above)
Comment (sanjula-rtc) [reply backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java:134]: valid

File: backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java (hunk 8 above)
Comment (sanjula-rtc) [reply backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java]: fixed

File: backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java (hunk 9 above)
Comment (sanjula-rtc) [reply backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java:134]: fixed

File: backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java (hunk 19 above)
Comment (sanjula-rtc) [reply backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java:178]: ordered

File: backend/src/test/java/com/skapp/community/crmplanner/controller/v1/CrmDealStageControllerIntegrationTest.java (hunk 18 above)
Comment (sanjula-rtc) [reply backend/src/test/java/com/skapp/community/crmplanner/controller/v1/CrmDealStageControllerIntegrationTest.java:10]: added

File: backend/src/main/java/com/skapp/community/crmplanner/payload/request/CrmDealStageEditRequestDto.java (hunk 20 above)
Comment (sanjula-rtc) [reply backend/src/main/java/com/skapp/community/crmplanner/payload/request/CrmDealStageEditRequestDto.java:1]: this is the practice

File: backend/src/main/java/com/skapp/community/crmplanner/controller/v1/CrmDealStageController.java (hunk 21 above)
Comment (sanjula-rtc) [reply backend/src/main/java/com/skapp/community/crmplanner/controller/v1/CrmDealStageController.java]: done simplifeid

**Review by yuhanga-rtc — DISMISSED** [review]

File: backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java (hunk 16 above)
Comment (sanjula-rtc) [reply backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java]: done

File: backend/src/main/java/com/skapp/community/crmplanner/repository/CrmDealStageDao.java (hunk 12 above)
Comment (sanjula-rtc) [reply backend/src/main/java/com/skapp/community/crmplanner/repository/CrmDealStageDao.java:4]: done

File: backend/src/main/java/com/skapp/community/crmplanner/constant/CrmMessageConstant.java (hunk 10 above)
Comment (sanjula-rtc) [reply backend/src/main/java/com/skapp/community/crmplanner/constant/CrmMessageConstant.java]: done

File: backend/src/main/java/com/skapp/community/crmplanner/util/CrmValidations.java (hunk 11 above)
Comment (sanjula-rtc) [reply backend/src/main/java/com/skapp/community/crmplanner/util/CrmValidations.java]: added invalid request

File: backend/src/main/java/com/skapp/community/crmplanner/constant/CrmMessageConstant.java (hunk 13 above)
Comment (sanjula-rtc) [reply backend/src/main/java/com/skapp/community/crmplanner/constant/CrmMessageConstant.java]: done

File: backend/src/main/java/com/skapp/community/crmplanner/constant/CrmMessageConstant.java (hunk 17 above)
Comment (sanjula-rtc) [reply backend/src/main/java/com/skapp/community/crmplanner/constant/CrmMessageConstant.java]: done

File: backend/src/test/java/com/skapp/community/crmplanner/controller/v1/CrmDealStageControllerIntegrationTest.java (hunk 15 above)
Comment (sanjula-rtc) [reply backend/src/test/java/com/skapp/community/crmplanner/controller/v1/CrmDealStageControllerIntegrationTest.java:312]: done

File: backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java [hunk 22]
```diff
@@ -68,6 +82,111 @@ public ResponseEntityDto createDealStage(CrmDealStageCreateRequestDto requestDto
...
+				throw new ModuleException(CrmMessageConstant.CRM_ERROR_DEAL_STAGE_NOT_FOUND);
+			}
+
+			stage.setOrderIndex(newStage.getOrderIndex());
+		});
+
+		ensureFirstStageIsInitial(existingStages);
+		crmDealStageDao.saveAll(existingStages);
+
+		existingStages.sort(Comparator.comparing(CrmDealStage::getOrderIndex));
```
Comment (akilarootcode) [inline backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java]: remove

**Review by akilarootcode — CHANGES_REQUESTED** [review]

File: backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java (hunk 22 above)
Comment (sanjula-rtc) [reply backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java]: removed

File: backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java [hunk 23]
```diff
@@ -68,6 +82,109 @@ public ResponseEntityDto createDealStage(CrmDealStageCreateRequestDto requestDto
...
+
+		ensureFirstStageIsInitial(existingStages);
+		crmDealStageDao.saveAll(existingStages);
+
+		log.info("reorderDealStages: execution ended");
+
+		return new ResponseEntityDto(false, crmMapper.crmDealStagesToCrmDealStageResponseDtos(existingStages));
+	}
+
+	private void ensureFirstStageIsInitial(List<CrmDealStage> reorderedStages) {
```
Comment (yuhanga-rtc) [inline backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java]: Consider changing the function name according to the functionality

Comment (sanjula-rtc) [reply backend/src/main/java/com/skapp/community/crmplanner/service/impl/CrmDealStageServiceImpl.java]: done

**Review by akilarootcode — DISMISSED** [review]

**Review by yuhanga-rtc — DISMISSED** [review]

File: backend/src/main/java/com/skapp/community/crmplanner/constant/CrmConstants.java [hunk 24]
```diff
@@ -57,4 +58,9 @@ public class CrmConstants {
 
 	public static final int DOMAIN_MAX_LENGTH = 255;
 
+	public static final Set<CrmDealStageType> TERMINAL_STAGES = EnumSet.of(CrmDealStageType.WON, CrmDealStageType.LOST);
+
+	public static final Set<CrmDealStageType> NON_DELETEABLE_STAGES = EnumSet.of(CrmDealStageType.INITIAL,
```
Comment (ThinuwanW) [inline backend/src/main/java/com/skapp/community/crmplanner/constant/CrmConstants.java]: typo

**Review by ThinuwanW — CHANGES_REQUESTED** [review]

File: backend/src/main/java/com/skapp/community/crmplanner/constant/CrmConstants.java (hunk 24 above)
Comment (sanjula-rtc) [reply backend/src/main/java/com/skapp/community/crmplanner/constant/CrmConstants.java]: fixed
